### PR TITLE
test: parameter client unit test 

### LINKED ISSUE #527: [TEST] Jax Parameter Client
### What do you want to test?
Jax paramater client components

### Outline of test structure
* Unit tests
* Test components and hooks

### Definition of done
Passing checks, cover all hooks, edge cases considered

### Mandatory checklist before making a PR
* [ ] The success criteria laid down in “Definition of done” are met.
* [ ] Test code is documented - docstrings for methods and classes, static types for arguments.
* [ ] Documentation is updated - README, CONTRIBUTING, or other documentation.

### PR DESCRIPTION
**Pull request by RuanJohn**
## What?
Unit tests for Jax parameter client component. 
## Why?
Ongoing work to increase testing code coverage for Jax Mava redesign. 
## Extra
closes #527

### CODE REVIEW

File: tests/jax/components/building/parameter_client_test.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+# limitations under the License.
+
+"""Unit tests for parameter client components"""
+
+from types import SimpleNamespace
+
+import pytest
+
+from mava.components.jax.building.parameter_client import (
+    ExecutorParameterClient,
```
Comment (OmaymaMahjoub) [inline tests/jax/components/building/parameter_client_test.py:23]: Missing `BaseParameterClient` test

File: tests/jax/components/building/parameter_client_test.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    """
+
+    mock_builder = mock_builder_with_parameter_client
+    mock_builder.store.is_evaluator = False
+    exec_param_client = ExecutorParameterClient(
+        config=ExecutorParameterClientConfig(executor_parameter_update_period=100)
+    )
+    exec_param_client.on_building_executor_parameter_client(builder=mock_builder)
+
+    assert hasattr(mock_builder.store, "executor_parameter_client")
```
Comment (OmaymaMahjoub) [inline tests/jax/components/building/parameter_client_test.py:74]: I think this is unnecessary since in the next lines you  verify that `mock_builder.store.executor_parameter_client...` exists

File: tests/jax/components/building/parameter_client_test.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    assert mock_builder.store.executor_parameter_client._get_keys == [
```
Comment (OmaymaMahjoub) [inline tests/jax/components/building/parameter_client_test.py:75]: I don't know if this necessary but there are some missing verifications like: 
`mock_builder.store.executor_parameter_client._all_keys`
`mock_builder.store.executor_parameter_client._client`... (Check this file mava/systems/jax/parameter_client.py)

**Review by OmaymaMahjoub — CHANGES_REQUESTED** [review]
Looks good to me, just small changes!

File: tests/jax/components/building/parameter_client_test.py (hunk 3 above)
Comment (RuanJohn) [reply tests/jax/components/building/parameter_client_test.py:116]: Good catch! I initially thought they were not too important but it makes sense to include them.

File: tests/jax/components/building/parameter_client_test.py (hunk 1 above)
Comment (RuanJohn) [reply tests/jax/components/building/parameter_client_test.py:26]: I initially left it out since it is only there to create count parameters, but I have now included an extra mock class and test case for it. Better to test than to not test.

File: tests/jax/components/building/parameter_client_test.py (hunk 3 above)
Comment (OmaymaMahjoub) [reply tests/jax/components/building/parameter_client_test.py]: Sounds great at least the ones that are included as parameter when it initiate an instance from the class like ._all_keys... 🙌

**Review by OmaymaMahjoub — APPROVED** [review]
Looks good to me! Thank you @RuanJohn

File: tests/jax/components/building/parameter_client_test.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+        "networks-network_agent_0",
+        "networks-network_agent_1",
+        "networks-network_agent_2",
+        "trainer_steps",
+        "trainer_walltime",
+        "evaluator_steps",
+        "evaluator_episodes",
+        "executor_episodes",
+        "executor_steps",
+    ]
```
Comment (sash-a) [inline tests/jax/components/building/parameter_client_test.py:137]: I think it might be a bit better to do:

```
expected_keys = set([
        "evaluator_episodes",
        "evaluator_steps",
        "executor_episodes",
        "executor_steps",
        "networks-network_agent_0",
        "networks-network_agent_1",
        "networks-network_agent_2",
        "trainer_steps",
        "trainer_walltime",
    ]
)

assert all([key in expected_keys for key in mock_builder.store.executor_parameter_client._all_keys])
assert all([key in expected_keys for key in mock_builder.store.executor_parameter_client._get_keys])
```

Unless I'm missing something but it seems like `get_keys` and `all_keys` are the same, but in a different order. So this should do the same, but it's a bit shorter and order agnostic.

File: tests/jax/components/building/parameter_client_test.py [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+    mock_builder = mock_builder_with_parameter_client
+    mock_builder.store.is_evaluator = True
+    exec_param_client = ExecutorParameterClient(
+        config=ExecutorParameterClientConfig(executor_parameter_update_period=100)
+    )
+    exec_param_client.on_building_executor_parameter_client(builder=mock_builder)
+
+    assert hasattr(mock_builder.store, "executor_parameter_client")
+
+    assert mock_builder.store.executor_parameter_client._all_keys == [
```
Comment (sash-a) [inline tests/jax/components/building/parameter_client_test.py:180]: Everything below seems the same as above test. Do you need to retest it? I would imagine you're testing the same code, could you just remove it? If I'm wrong and it's different code generating the same data then feel free to ignore this

**Review by sash-a — CHANGES_REQUESTED** [review]
I think you could clean this up a lot if you just had one big method that tests some kind of parameter client ie `def check_parameter_client(parameter_client):`. Where you put all the asserts that you are repeating in three of the test cases. And then you can pass into this function either `mock_builder.store.executor_parameter_client` or `mock_builder.store.trainer_parameter_client` and it should do all the asserts.

But otherwise than this it looks good! :rocket:

File: tests/jax/components/building/parameter_client_test.py (hunk 4 above)
Comment (RuanJohn) [reply tests/jax/components/building/parameter_client_test.py:137]: Totally agree. This definitely cleans things up really well. I've done this for all test cases.

File: tests/jax/components/building/parameter_client_test.py (hunk 5 above)
Comment (RuanJohn) [reply tests/jax/components/building/parameter_client_test.py]: It is indeed different (although slightly) since the below case is for an executor to be used for evaluation and the above was for a 'vanilla' executor.

**Comment by RuanJohn** [comment]
> I think you could clean this up a lot if you just had one big method that tests some kind of parameter client ie `def check_parameter_client(parameter_client):`. Where you put all the asserts that you are repeating in three of the test cases. And then you can pass into this function either `mock_builder.store.executor_parameter_client` or `mock_builder.store.trainer_parameter_client` and it should do all the asserts.
> 
> But otherwise than this it looks good! rocket

Thanks for the feedback! You are right there are quite a few repeated asserts throughout the tests, but I think it still makes more sense to keep all the cases separate. There are 3 cases for the `ExecutorParameterClient` (executor with parameter client passed in, evaluator with a parameter client passed in and a case where no parameter client is passed in through the builder store) and there are 2 for the `TrainerParameterClient` (with and without a parameter client passed in through the builder store). 

It seems to me that having a big method would require multiple fixtures or multiple set ups inside the method which would add a bunch of code. If there is a cleaner way please let me know? Happy to change things :)

**Comment by sash-a** [comment]
> > I think you could clean this up a lot if you just had one big method that tests some kind of parameter client ie `def check_parameter_client(parameter_client):`. Where you put all the asserts that you are repeating in three of the test cases. And then you can pass into this function either `mock_builder.store.executor_parameter_client` or `mock_builder.store.trainer_parameter_client` and it should do all the asserts.
> > But otherwise than this it looks good! rocket
> 
> Thanks for the feedback! You are right there are quite a few repeated asserts throughout the tests, but I think it still makes more sense to keep all the cases separate. There are 3 cases for the `ExecutorParameterClient` (executor with parameter client passed in, evaluator with a parameter client passed in and a case where no parameter client is passed in through the builder store) and there are 2 for the `TrainerParameterClient` (with and without a parameter client passed in through the builder store).
> 
> It seems to me that having a big method would require multiple fixtures or multiple set ups inside the method which would add a bunch of code. If there is a cleaner way please let me know? Happy to change 

I think the change you made with extracting the expected variables has cleaned this up a whole lot. I'm happy to keep it the way it is :rocket:

**Review by sash-a — APPROVED** [review]
:1st_place_medal:

**Review by OmaymaMahjoub — APPROVED** [review]
